Add troubleshooting tips to prevent ProGuard over optimization (#28)
cossacklabs/product-docs: themis/languages/java/installation-android.md

# Installing JavaThemis for Android development

> JavaThemis for Android is [available on JFrog Bintray via **JCenter** and **Maven**](https://bintray.com/cossacklabs/maven/themis).

Reviewer: Maven is a tool. Bintray and JCenter are repositories.
```suggestion
JavaThemis for Android is [available on JCenter and Bintray via **Maven**](https://bintray.com/cossacklabs/maven/themis).
```

Author: thank you, I'm being constantly confused by this

Once JavaThemis is installed, you can [try out examples on your machine](../examples/).

## Configuring ProGuard / R8 rules

> Sometimes when working with Android projects in `Release` configuration, you might get runtime error, like:

Reviewer: This section is about a particular runtime error, not about dealing with runtime errors in general. I'd rephrase the sentence a bit to suggest that what follows is an example. Also, clauses usually don't need commas.

```suggestion
Sometimes when working with Android projects in `Release` configuration you might get the following runtime error:
```

```
Java.Lang.IncompatibleClassChangeError: no non-static method "Lcom/cossacklabs/themis/SecureCellSeal;.decrypt([B[B)[B"
[orion.mobile]   at Java.Interop.JniEnvironment+InstanceMethods.GetMethodID (Java.Interop.JniObjectReference type, System.String name, System.String signature) [0x0005b] in <42d2b7086f0a46efb99253c5db1ecca9>:0 
[orion.mobile]   at Android.Runtime.JNIEnv.GetMethodID (System.IntPtr kls, System.String name, System.String signature) [0x00007] in <3080427739614e60a939a88bf3f838d5>:0 
[orion.mobile]   at Com.Cossacklabs.Themis.SecureCell+ISealInvoker.Decrypt (System.Byte[] p0, System.Byte[] p1) [0x00017] in <cd618986d1ce4194b63cdd3366dad291>:0 
[orion.mobile]   at Themis.Droid.CellSealDroid.UnwrapData (Themis.ISecureCellData cipherTextData, System.Byte[] context) [0x0007e] in <a492e7118e094c3296442a386fe5d80e>:0 
[orion.mobile]    --- End of inner exception stack trace --
```

> It is caused by "over-optimizing" a build by ProGuard / R8, which strips native Java classes from a build. 

Reviewer: Rephrased a bit so that the "blame" falls on ProGuard, rather then the developers who "over-optimized their build".

```suggestion
It is caused by ProGuard / R8 "over-optimizing" a build, stripping native Java classes as a result.
```

> An easy solution is to add `proguard.cfg` file to the Android app project with the following lines:

Reviewer: A bit of a personal touch:
```suggestion
An easy solution is to add `proguard.cfg` file to your Android app project with the following lines:
```

Author: As I understood, there're certain Proguard "default" configurations on Release mode, which users might tweak with additional settings.

```
-keep class com.cossacklabs.themis.**
-keep class com.cossacklabs.themis.** {*;}
```

> Make sure that the project has a `ProguardConfiguration` build step.

Reviewer: A bit more...
```suggestion
Make sure to enable the `ProguardConfiguration` build step after that.
```

> Please refer to [the issue #716](https://github.com/cossacklabs/themis/issues/716) for details, and to the [example application made by @dodikk](https://github.com/dodikk/themis-xamarin-prototype/pull/10) who has discovered and troubleshoot the issue.

Reviewer: I believe this verb should be inflected like this here:
```suggestion
Please refer to [the issue #716](https://github.com/cossacklabs/themis/issues/716) for details, and to the [example application made by @dodikk](https://github.com/dodikk/themis-xamarin-prototype/pull/10) who has discovered and troubleshot the issue.
```
